core/ledger/kvledger/channelinfo_provider.go

```diff
@@ -0,0 +1,69 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package kvledger
+
+import (
+	"github.com/hyperledger/fabric/bccsp/factory"
+	"github.com/hyperledger/fabric/common/channelconfig"
+	"github.com/hyperledger/fabric/common/ledger/blkstorage"
+	"github.com/hyperledger/fabric/protoutil"
+)
+
+// ChannelInfoProvider implements ledger.ChannelInfoProvider interface
+type ChannelInfoProvider struct {
+	BlkStoreProvider *blkstorage.BlockStoreProvider
+}
+
+// GetAllMSPIDs retrieves the MSP IDs of the organizations in all channel configurations (current and previous
+// channel configurations). It returns a mapping of org names to MSPIDs.
+func (ciProvider *ChannelInfoProvider) GetAllMSPIDs(channelName string) (map[string]string, error) {
+	blockStore, err := ciProvider.BlkStoreProvider.Open(channelName)
+	if err != nil {
+		return nil, err
+	}
+	blockchainInfo, err := blockStore.GetBlockchainInfo()
+	if err != nil {
+		return nil, err
+	}
+
+	// Iterate over the config blocks to retrieve all the channel configs and construct orgs to MSPIDs mapping
+	blockNum := blockchainInfo.Height
+	orgsToMSPIDs := map[string]string{}
+	for blockNum > 0 {
+		channelConfig, configBlockNum, err := mostRecentChannelConfigBelow(blockNum, blockStore)
+		if err != nil {
+			return nil, err
+		}
+
+		orgs := channelConfig.ApplicationConfig().Organizations()
+		for org, appOrg := range orgs {
+			if _, ok := orgsToMSPIDs[org]; !ok {
+				orgsToMSPIDs[org] = appOrg.MSPID()
+			}
+		}
+		blockNum = configBlockNum
+	}
+
+	return orgsToMSPIDs, nil
+}
+
+// mostRecentChannelConfigBelow fetches the most recent config block below the blockNum and extracts the channel config from it.
+func mostRecentChannelConfigBelow(blockNum uint64, blockStore *blkstorage.BlockStore) (*channelconfig.ChannelConfig, uint64, error) {
+	block, err := blockStore.RetrieveBlockByNumber(blockNum - 1)
+	configBlockIndex, err := protoutil.GetLastConfigIndexFromBlock(block)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	configBlock, err := blockStore.RetrieveBlockByNumber(configBlockIndex)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	channelConfig, err := channelconfig.ExtractChannelConfig(configBlock, factory.GetDefault())
+	return channelConfig, configBlockIndex, err
```

Comment:
In this last case shouldn't u return nil for `err`? 
```go
return channelConfig, configBlockIndex, nil
```

Reply:
It should return err which may be nil or a real error.

Reply:
It should return err which may be nil or a real error. I updated the code to be more readable.